### PR TITLE
Constrain sdpa to fx strides

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112045
* __->__ #111721
* #111976

Fix for https://github.com/pytorch/pytorch/issues/109607. sdpa requires last dimension strides to be 1. Add constraint so that we run the op with the strides we observed in tracing.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by drisspg — APPROVED** [review]
👍

**Comment by Chillee** [comment]
Ah I ran into this too at some point.

**Review by Chillee — APPROVED** [review]

**Review by jansel — APPROVED** [review]

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -2001,10 +2001,71 @@ def apply_constraint(arg, fx_arg):
...
+        if not isinstance(arg, ir.IRNode):
+            return arg
+
+        meta_val = fx_arg.meta["val"]
+        if not meta_val.is_cuda:
+            return arg
+
+        stride_order = ir.get_stride_order(meta_val.stride())
+        if stride_order and stride_order[-1] != 0:
+            stride_order = list(reversed(range(len(arg.get_size()))))
```
Comment (drisspg) [inline torch/_inductor/lowering.py:2018]: I know what this does but maybe a comment saying "constructs contiguous stride order

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -2001,10 +2001,71 @@ def apply_constraint(arg, fx_arg):
...
+
+        efficient_kernels = (
+            aten._scaled_dot_product_efficient_attention.default,
+            aten._scaled_dot_product_efficient_attention_backward.default,
+        )
+        ALIGNMENT = 16 if fx_node.target in efficient_kernels else 8
```
Comment (drisspg) [inline torch/_inductor/lowering.py:2024]: why not only 16?

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -2001,10 +2001,71 @@ def apply_constraint(arg, fx_arg):
...
+        aligned_input_view = unaligned_input_shape and is_aligned(arg.unwrap_view())
+
+        # input is padded, requiring_stride_order will unwrap the view and unpad.
+        # Would be nice to be able to require certain padding from inductor ir, nyi
+        if aligned_input_view:
+            return arg
+
+        return ir.ExternKernel.require_stride_order(arg, stride_order)
+
+    args = tuple(
```
Comment (drisspg) [inline torch/_inductor/lowering.py:2041]: nit: would some pytree  func here be prettier?

**Review by drisspg — APPROVED** [review]
Left some nit comments but looks good module the failing test

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/lowering.py:2024]: correct me if i'm wrong here. i was getting an error for 8 alignment in cutlass. should it be 16 ?

Comment (drisspg) [reply torch/_inductor/lowering.py:2024]: https://github.com/pytorch/pytorch/blob/b70efde3ad6ee7385eab9fa3d40ae45236c9e4a3/aten/src/ATen/native/transformers/attention.cpp#L544

I have been using 16 which I think is the requirement

File: torch/_inductor/lowering.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/lowering.py:2038]: Pytree doesn't preserve the key-value pairing which we need here.

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but it is breaking ROCm job in trunk https://hud.pytorch.org/pytorch/pytorch/commit/8a7c3cec78686e661b3781b916a8aae59083f90a' -c nosignal

Note that ROCm jobs have been move to a separate workflow, you can add`ciflow/rocm` to run it on your PR.

**Pull request closed without merging**